### ISSUE TITLE
Massive slowdown when updating from 2.5.1 to 2.6.0

### ISSUE THREAD

**Issue by Trainmaster** [squizlabs/PHP_CodeSniffer#952]
In our company we observed a massive slowdown when running CodeSniffer 2.6.0 with a custom ruleset:

Version | Small project (579 files) | Large project (5486 files)
--- | --- | ---
2.5.1 | Time: 12.89 secs; Memory: 46Mb | Time: 3 mins, 12.88 secs; Memory: 56Mb
2.6.0 | Time: 47.68 secs; Memory: 46Mb | Time: 8 mins, 34.95 secs; Memory: 56Mb

Both analyses were run with PHP7:
```bash
$ php -v
PHP 7.0.5 (cli) (built: Mar 30 2016 10:15:34) ( NTS )
Copyright (c) 1997-2016 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2016 Zend Technologies
```

Has anybody else experienced such a slowdown?

**Comment by aik099** [comment]
I bet this is due change in the indentation checking sniff where it now reports an error for each incorrectly indented line in a scope in contrast to where it was reporting error on first line only.

**Comment by montchr** [comment]
I've been having this problem as well.

**Comment by gsherwood** [comment]
I haven't experienced any slowdown during development, and the travis runs haven't seen a decrease in performance either.

So the first question is @Trainmaster or @montchr: any chance either your repos or standards are open source and I can look?

If not, the way to get to the bottom of it is to run 2.5.1 and 2.6.0 with the -v CLI argument. This will output the number of errors and processing time for every file. By looking at this output for both versions, we can tell if it is a specific file (or set of files) or an overall slowdown you are seeing.

I'd also recommend printing a summary report (--report=summary) so that we can see a general error count and check if more errors are being produced.

So if your command is currently `phpcs --standard=/path/to/standard /path/to/code` please change it to `phpcs --standard=/path/to/standard /path/to/code --report=summary -v` and compare the output.

If anyone can get me a specific file that is showing a really obviously performance decrease, that might be enough to find a solution.

Thanks for any help you can provide.

**Comment by Trainmaster** [comment]
@gsherwood I think I've tracked down the rule causing the decrease in performance, it's`Generic.PHP.Syntax`. 

With `2.5.1` each file takes about 0-31 ms, after updating to `2.6.0` 80-109 ms. After excluding the rule `Generic.PHP.Syntax` the difference is gone and `2.6.0` also shows 0-31 ms.

**Comment by aik099** [comment]
If I remember correctly, then this rule is just calling `php -l` on a file. Not sure how that can be slower.

What you if you create ruleset just from that one sniff and run it. Would it be slower as well?

**Comment by Bufallo** [comment]
I can confirm the massive performance decrease.

Im using PSR2 Sniffs on 107 php-files.
Also im using grunt phpcs for running the checks.

With 2.5.1 phpcs needs 2.2 seconds (according to time-grunt).
With 2.6.0 phpcs needs 62,5 seconds.

PHP Version: PHP 5.6.20 (cli)

**Comment by Trainmaster** [comment]
@aik099 

I tested it with the following ruleset:
```xml
<?xml version="1.0"?>
<ruleset>
    <rule ref="Generic.PHP.Syntax"/>
</ruleset>
```
and it's slower as well (14.56 secs vs 3.45 secs).

**Comment by gsherwood** [comment]
The only change to that sniff is the auto-detection of the PHP binary if you haven't set one.

So what I suspect you are seeing is that the syntax check was not previously being run because you didn't set the php_path config option (correct me if I'm wrong - it's just a guess) but now this is being auto-detected and so the check is running. The extra time is obviously how long it takes to call out to PHP and check if the file syntax is correct.

So previously, you might have included this sniff but it was silently not doing anything.

An easy way to test on PHPCS 2.5.1 is to add `--runtime-set php_path /path/to/php` to your command. This will ensure that the syntax check actually runs.

**Comment by Trainmaster** [comment]
Well, it looks like `$phpPath` in `SyntaxSniff.php:57` is null and therefore doesn't work in `2.5.1`. But in `2.6.0` a fallback (`PHP_BINARY`) was added for PHP 5.4+.

**Comment by aik099** [comment]
😄

**Comment by gsherwood** [comment]
Glad we figured this one out. You can make a decision as to if you want to include the syntax check if your runs or not now based on this performance info.
